### PR TITLE
Update vamp signatures (replaces old PR#208)

### PR DESCRIPTION
**Pull request by upside2**
Add signatures for ntdll security_check_cookie from Visual Studio versions thru 2019.  The 64-bit 2019 signature appears in unit test tests/testvivisect.py.

Add signatures for ntdll seh4_prolog and seh4_epilog from Visual Studio versions thru 2019.

### CODE REVIEW

**Comment by atlas0fd00m** [comment]
cool, thanks @upside2 !

would you mind adding a unittest for these so that we don't accidentally break them?
it can even be a made-up workspace where you add the necessary bytes, call `vw.makeFunction(va)`, and check that it identified the right thing. 

thanks!
@

**Comment by upside2** [comment]
Added a simple test as you described.

File: vivisect/tests/testvampsigs.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        opcodes = '68aaaaaaaa64a10000000050'
+        assert self.run_test(opcodes) == 'ntdll.seh3_prolog'
+
+        # Signature with extra opcodes.
+        opcodes = '680000000064a100000000504141414141'
+        assert self.run_test(opcodes) == 'ntdll.seh3_prolog'
+
+        # Signature with opcodes before.
+        opcodes = '4141680000000064a10000000050'
+        assert self.run_test(opcodes) is None
```
Comment (atlas0fd00m) [inline vivisect/tests/testvampsigs.py:61]:
```suggestion
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh3_prolog')

        # Incorrect signature.
        opcodes = '680000000064a100000000aa'
        self.assertEqual(self.run_test(opcodes), None)

        # Signature with masked out bytes changed.
        opcodes = '68aaaaaaaa64a10000000050'
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh3_prolog')

        # Signature with extra opcodes.
        opcodes = '680000000064a100000000504141414141'
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh3_prolog')

        # Signature with opcodes before.
        opcodes = '4141680000000064a10000000050'
        self.assertEqual(self.run_test(opcodes), None)
```

Comment (atlas0fd00m) [reply vivisect/tests/testvampsigs.py:61]: for unittests, we want to use `Unittest.assert*` functions

File: vivisect/tests/testvampsigs.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        assert self.run_test(opcodes) == 'ntdll._alloca_probe'
+
+        opcodes = '518d4c24042bc81bc0f7d023c88bc42500f0ffff3bc8720a8bc159948b00890424c32d001000008500ebe9'
+        assert self.run_test(opcodes) == 'ntdll._alloca_probe'
+
+        opcodes = '518d4c24042bc81bc0f7d023c88bc42500f0ffff3bc8f2720b8bc159948b00890424f2c32d001000008500ebe7'
+        assert self.run_test(opcodes) == 'ntdll._alloca_probe'
+
+        opcodes = '4883ec104c8914244c895c24084d33db4c8d5424184c2bd04d0f42d3654c8b1c25100000004d3bd37316664181e200f04d8d9b00f0ffff41c603004d3bd375f04c8b14244c8b5c24084883c410c3'
+        assert self.run_test(opcodes) == 'ntdll._alloca_probe'
```
Comment (atlas0fd00m) [inline vivisect/tests/testvampsigs.py:101]:
```suggestion
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh3_prolog')

        opcodes = '8b4df064890d00000000595f5e5bc951c3'
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh3_epilog')

        opcodes = '680000000064ff35000000008b442410'
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh4_prolog')

        opcodes = '8b4df064890d00000000595f5f5e5b8be55d51c3'
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh4_epilog')

        opcodes = '8b4df064890d00000000595f5f5e5b8be55d51f2c3'
        self.assertEqual(self.run_test(opcodes), 'ntdll.seh4_epilog')

        opcodes = 'a10000000033c58945fc'
        self.assertEqual(self.run_test(opcodes), 'ntdll.gs_prolog')


    def test_alloca_probe_sigs(self):
        '''
        Test detection of alloca_probe signatures.
        '''
        opcodes = '513d001000008d4c2408721481e9001000002d0010000085013d0010000073ec2bc88bc485018be18b088b400450c3'
        self.assertEqual(self.run_test(opcodes), 'ntdll._alloca_probe')

        opcodes = '518d4c24042bc81bc0f7d023c88bc42500f0ffff3bc8720a8bc159948b00890424c32d001000008500ebe9'
        self.assertEqual(self.run_test(opcodes), 'ntdll._alloca_probe')

        opcodes = '518d4c24042bc81bc0f7d023c88bc42500f0ffff3bc8f2720b8bc159948b00890424f2c32d001000008500ebe7'
        self.assertEqual(self.run_test(opcodes), 'ntdll._alloca_probe')

        opcodes = '4883ec104c8914244c895c24084d33db4c8d5424184c2bd04d0f42d3654c8b1c25100000004d3bd37316664181e200f04d8d9b00f0ffff41c603004d3bd375f04c8b14244c8b5c24084883c410c3'
        self.assertEqual(self.run_test(opcodes), 'ntdll._alloca_probe')
```

File: vivisect/tests/testvampsigs.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        '''
+        Test detection of gs_failure signatures.
+        '''
+        opcodes = '3b0d000000007502f3c3e9'
+        assert self.run_test(opcodes) == 'ntdll.security_check_cookie'
+        opcodes = '8bff558bec5151a300000000890d00000000891500000000891d00000000893500000000893d000000008c15000000008c0d000000008c1d000000008c05000000008c25000000008c2d000000009c'
+        assert self.run_test(opcodes) == 'ntdll.report_gsfailure'
+
+        opcodes = '8bff558bec81ec28030000a300000000890d00000000891500000000891d00000000893500000000893d00000000668c1500000000668c0d00000000668c1d00000000668c0500000000668c2500000000668c2d000000009c'
+        assert self.run_test(opcodes) == 'ntdll.report_gsfailure'
```
Comment (atlas0fd00m) [inline vivisect/tests/testvampsigs.py:137]:
```suggestion
        self.assertEqual(self.run_test(opcodes), 'ntdll._alloca_probe')


    def test_security_cookie_sigs(self):
        '''
        Test detection of security_check_cookie signatures.
        '''
        opcodes = '3b0d000000007502f3c3e9'
        self.assertEqual(self.run_test(opcodes), 'ntdll.security_check_cookie')

        opcodes = '3b0d00000000f27502f2c3f2e9'
        self.assertEqual(self.run_test(opcodes), 'ntdll.security_check_cookie')

        opcodes = '483b0d00000000751148c1c11066f7c1ffff7502f3c348c1c910e9'
        self.assertEqual(self.run_test(opcodes), 'ntdll.security_check_cookie_64')

        opcodes = '483b0d00000000f2751148c1c11066f7c1fffff27502f2c348c1c910e9'
        self.assertEqual(self.run_test(opcodes), 'ntdll.security_check_cookie_64')

        opcodes = '483b0d00000000f2751248c1c11066f7c1fffff27502f2c348c1c910e9'
        self.assertEqual(self.run_test(opcodes), 'ntdll.security_check_cookie_64')


    def test_gs_failure_sigs(self):
        '''
        Test detection of gs_failure signatures.
        '''
        opcodes = '3b0d000000007502f3c3e9'
        self.assertEqual(self.run_test(opcodes), 'ntdll.security_check_cookie')
        opcodes = '8bff558bec5151a300000000890d00000000891500000000891d00000000893500000000893d000000008c15000000008c0d000000008c1d000000008c05000000008c25000000008c2d000000009c'
        self.assertEqual(self.run_test(opcodes), 'ntdll.report_gsfailure')

        opcodes = '8bff558bec81ec28030000a300000000890d00000000891500000000891d00000000893500000000893d00000000668c1500000000668c0d00000000668c1d00000000668c0500000000668c2500000000668c2d000000009c'
        self.assertEqual(self.run_test(opcodes), 'ntdll.report_gsfailure')
```

**Comment by atlas0fd00m** [comment]
thanks, @upside2 !  so close to being done.  we want to use Unittest's assertion functions instead of straight up `assert`.  i made the suggested changes for you.  i don't think you'll mind, so i'll just commit those suggestions (since it looks like i have that ability on this PR) and if the unittests all pass, i'll merge it in.

you're replacing #208 with this, but 208 had a bunch of other goodness in it.  might we see another PR for updating ImpApi stuffs?

**Review by atlas0fd00m — APPROVED** [review]
thanks @upside2 !

**Comment by upside2** [comment]
Your assertEquals changes are fine.  Looks good, thanks.

**Pull request merged**